## Intro

Upon completing Phase 2 of this project, where the initial processed dataset was uploaded to a MySQL database and reviewed with SQL querying, with duplicate rows being deleted, we are now ready to address the issue pertaining to each property's designated district. For a background discussion of this issue, refer back to report published at: https://reggiealderson.github.io/portfolio/blog/nswpropertydataproject/

In this phase we will be utilising the NSW Government's Address Location Web Service API, accessible via: http://maps.six.nsw.gov.au/sws/AddressLocation.html

The API, if called correctly, will return geographical information (stored in a JSON object) for each property we wish to retrieve information for. 



## Loading the dataset from the SQL database and saving to a new python dataframe

In [103]:
#Importing the relevant modules

import pandas as pd
import sqlalchemy
import os, csv, re
import urllib.parse
import urllib.request
import urllib.error
import json
from datetime import datetime

In [104]:
engine = sqlalchemy.create_engine('mysql+pymysql://root:learningSQL@localhost:3306/testdatabase') #This allows python to interact with your MySQL database. # Tutorial to be watched to understand this line of code: https://www.youtube.com/watch?v=M-4EpNdlSuY

In [105]:
dataset = pd.read_sql_table("property_sales", engine) #This reads the dataset from the property_sales table in the MySQL database, and saves a copy as a pandas dataframe

In [106]:
#Checking the dataset was read correctly from the MySQL table

print('Number of rows, number of columns: ')
print(dataset.shape)
print('\n')
print('First five rows of the dataset: ')
dataset.head()

Number of rows, number of columns: 
(28651, 16)


First five rows of the dataset: 


district_code  property_id property_unit_number property_house_number  \
0              1    4163208.0                 None                     1   
1              1       4037.0                 None                   332   
2              1       2170.0                 None                    67   
3              1       6554.0                 None                     1   
4              1       6621.0                 None                    63   

  property_street_name property_locality  property_post_code     area  \
0        ST ANDREWS CL      HEDDON GRETA                2321   782.10   
1          WOLLOMBI RD  BELLBIRD HEIGHTS                2325   809.40   
2           CHARLES ST          ABERMAIN                2326  6373.00   
3             EDITH ST          CESSNOCK                2325   809.43   
4          FERGUSON ST          CESSNOCK                2325   696.77   

  contract_date settlement_date  purchase_price zoning_classification  \
0    2020-01-14      2020-02-25          240000                    R2   
1    2020-01-28      2020-02-27          275000                    R2   
2    2020-02-05      2020-02-21          200000                    R5   
3    2020-01-24      2020-02-24          290000                    R3   
4    2020-01-29      2020-02-26          300000                    R3   

   is_unit  is_house record_id  id  
0      NaN       1.0  AP924668   1  
1      NaN       1.0  AP927983   2  
2      NaN       1.0  AP917863   3  
3      NaN       1.0  AP921667   4  
4      NaN       1.0  AP926464   5

## Data transformation (prior to use of API)

In this section we will loop through each property transaction (row in the dataframe), and derive a new column which contains a clean string of the street name a property belongs to. This is necessary for a couple of reasons. Firstly, the raw dataset is fairly inconsistent with the structure of the street name data point across all of the property transactions. For instance, some records will mistakenly have numbers and leading white spacesd in the street name, while others may contain duplicate road type strings (e.g. 'CHARLES ROAD RD'). Secondly, the API we will be using requires that we enter a road name parameter, which only accepts the road name (not including the road type). 

In the below cell, I have referenced two csv files which contain road name types and abbreviations (e.g. 'RD', 'CRES', 'STREET', 'AVENUE', 'LNWY'). If these types and abbreviations are included in the road name parameter when calling the API, then the returned JSON object will contain no information. The csv files are available here: https://github.com/reggiealderson/NSWpropertysales . The choice as to which words and abbreviations to include in these csv files was difficult. For instance, do I include a word like 'TERRACE' which could be both a road type and also a legitimate word in a street name (e.g. RAYMOND TERRACE ROAD). I had to do several trial runs with the API, checking the dataset with SQL queries to see the prevalence of certain abbreviations and road type words in the dataset, and then updating the csv lists, before I could be happy with the % of missing values generated from the API call. 

In [107]:
# Setting up python lists that will be eventually looped through for each property record

abbrevfile_obj = open('abbrev.csv')
typesfile_obj = open('types.csv')
abbrevreader_obj = csv.reader(abbrevfile_obj)
typesreader_obj = csv.reader(typesfile_obj)

listofabbrev = list(abbrevreader_obj)
listofabbrev[0][0] = 'ACCS' #This is just to fix up the first entry as for some reason it gets read wrong by the csv reader. 
listofabbrev = listofabbrev[0] #Refers to the first row in the csv file.

listoftypes = list(typesreader_obj)
listoftypes[0][0] = 'ACCESS' #This is just to fix up the first entry as for some reason it gets read wrong by the csv reader. 
listoftypes = listoftypes[0] #Refers to the first row in the csv file.

abbrevfile_obj.close()
typesfile_obj.close()

In [108]:
cols = [15,3,4,5] # Refering to the columns in the dataset that we need information for. 

In [129]:
properties_list = dataset[dataset.columns[cols]] # (1 of 2) Creates a python list containing the property data we will be transforming and then subsequently passing through the API

In [130]:
properties_list = properties_list.values.tolist() # (2 of 2) Creates a python list containing the property data we will be transforming and then subsequently passing through the API

In [131]:
properties_list[17974] # A random check of property data (use as reference)

[22062, None, 'AERODROME ROAD RD', 'LIGHTNING RIDGE']

In [155]:
# Removing suffixes such as 'ST', 'RD', etc from the street name strings -- so that the future API calls can function properly
# This cell should take around 5 seconds to process.

for i in properties_list:
    counter = 0
    sec_counter = 0
    stn1 = i[2] # Refers to the street name
    stn2 = ''.join([i for i in stn1 if not i.isdigit()]).lstrip() # Gets rid of digits and leading white spaces
    stn3 = re.sub(r'(?:^| )\w(?:$| )', ' ', stn2).strip() # Gets rid of singular characters in the string 
    stn4 = stn3 + ' ' 
    for idx, abbrev in enumerate(listofabbrev):
        ab1 = ' ' + abbrev + ' ' # creates a string for each possible abbreviation, with white spaces on either side so that when searching for the string it is not confused with a word which contains the abbreviation (e.g. the 'ST' abbreviation features in the word 'STUART')
        stn5 = re.search(f"{ab1}$", stn4) # Here we are searching for the abbreviation in the property's street name
        if(stn5!=None): # If this IF condition is met, then the abbreviation features in the property's street name and a couple of variables are created (stored in memory only)
            counter += 1
            next_idx_pos = idx + 1
            ab2 = ab1 
            ab3 =  ' ' + abbrev
            f_abbrev = abbrev
            if stn3 == 'THE' + ab3:
                i.append(stn3)
            else: # For all other circumstances where the abbreviation is featured in the street name, it is removed and a new variable is added to the properties_list.
                ab4 = ' ' + f_abbrev + ' ' + f_abbrev
                stn6 = re.search(f"{ab4}$", stn3)
                if(stn6!=None): #If there is a repeated abbreviation (e.g 'RD RD')
                    var1 = stn3.replace(ab4, '')
                    i.append(var1)
                else:
                    var2 = stn4.replace(ab2, '')
                    for abbrev2 in listofabbrev[next_idx_pos:]: #checks if there is a second distinct abbreviation
                        ab1a = ' ' + abbrev2
                        stn5a = re.search(f"{ab1a}$", var2)
                        if(stn5a!=None): # if there is a second distinct abbreviation
                            var3 = var2.replace(ab1a, '')
                            i.append(var3)
                            sec_counter += 1
                        else:
                            continue
                    if sec_counter == 0: # if there is not a second distinct abbreviation, it falls back to var2
                        i.append(var2)
                    else:
                        continue
        else:
            continue

    if counter == 0: #if there were no abbreviations then a new variable is still added to the properties_list, it is just a copy of the original street name string. 
        i.append(stn3)


In [133]:
# Running this cell you will now see that a new data point is added for each property in properties_list. The new data point represents a semi cleaned street name string. However we still need to create a script that will handle strings such as 'ROAD', 'AVENUE', etc.

properties_list[17974]
# properties_list[108]

[22062, None, 'AERODROME ROAD RD', 'LIGHTNING RIDGE', 'AERODROME ROAD']

In [134]:
properties_list[3900] # Another random check of property data (use as reference)

[3901, '2', 'BELLCAST RD BUILDING A', 'ROUSE HILL', 'BELLCAST']

In [156]:
#Removing instances where the street name string ends in a substring like "ROAD" or "STREET" -- so that the future API calls can functions properly

for i in properties_list:
    tt1 = i[4] #refers to the previously created string holding the semi cleaned street name
    occureda = 0 
    occuredb = 0 
    occuredc = 0 
    for typ in listoftypes:
        tt2 = ' ' + typ
        tt3 = re.search(f"{tt2}$", tt1)
        if(tt3!=None):
            for typ2 in listoftypes:
                pp1 = typ2 + tt2
                if tt1 == pp1:
                    i.append(tt1)
                    occureda += 1
                else:
                    continue
            if tt1.startswith("THE "): #This condition is here in case the street name starts with 'The'. For instance if the street name is "THE GREAT OCEAN ROAD", then the API will actually need the 'ROAD' part of the string to be entered into the street name parameter. Whereas in all other cases (e.g. 'BAKERS ROAD') the API requires the 'ROAD' part to be excluded in the parameter.
                i.append(tt1)
                occuredb += 1 
            else:
                rr1 = tt1.replace(tt2, '')
                i.append(rr1)
                occuredc += 1 
        else:
            continue
    if ((occureda == 0) and (occuredb == 0) and (occuredc == 0)): 
        i.append(tt1)
    else:
        continue

In [136]:
# A random check of property data (use as reference)
properties_list[108]

[109, None, 'WILLIAM ST ST', 'PAXTON', 'WILLIAM', 'WILLIAM']

In [137]:
properties_list[17974] # A random check of property data (use as reference)

[22062,
 None,
 'AERODROME ROAD RD',
 'LIGHTNING RIDGE',
 'AERODROME ROAD',
 'AERODROME']

In [138]:
# In this example, 'ROAD' is kept in the final cleaned street name variable - because the street name starts with 'THE' - making it a special case.
properties_list[1066]

[1067,
 '24',
 'THE RIDGE ROAD',
 'EAST MAITLAND',
 'THE RIDGE ROAD',
 'THE RIDGE ROAD']

## Interacting with the geo-coding API

As mentioned, we will be using the NSW Government's Address Location Web Service API in order to gain certain geographical data for each property we pass through to the API. In our case, we wish to retrieve a property's designated district name, and it's longitude and latitude coordinates. 

In [139]:
len(properties_list) # Checking we have the correct amount of properties

28651

In [119]:
# I recommend doing a test run of the API with a much smaller dataset. Here I've created a list 'testrun_list' which you can replace with 'properties_list' in the next cell once you are satisfied that the API call will run smoothly.

testrun_list = properties_list[0:150] 

### The actual API call

The next cell runs the API call. Note this took my PC roughly 45 minutes to process with a sample of 28,651 properties. 

In [157]:
start_time = datetime.now()
BASE_URL = 'http://maps.six.nsw.gov.au/services/public/Address_Location'
current_delay = 0.1  # Set the initial retry delay to 100ms.
max_delay = 5  # Set the maximum retry delay to 5 seconds.
count_of_failed_data_retrieval = 0
count_of_failed_urlaccess = 0

for i in properties_list: # Here is where you substitute your list if you are looking to do a test run with a smaller dataset
    roadName = i[5].replace(" ", "+")
    suburb = i[3].replace(" ", "+")
    if i[1] != '':
        houseNumber_uncleaned = i[1]
        houseNumber_uncleaned_str = str(houseNumber_uncleaned)
        housenumber_numericonly = re.sub("[^0-9]", "", houseNumber_uncleaned_str)
        houseNumber = housenumber_numericonly.replace(" ", "+")
    else:
        houseNumber = ''
        
    url = f"{BASE_URL}?houseNumber={houseNumber}&roadName={roadName}&suburb={suburb}&projection=EPSG%3A4326"
    while True:
        try:
            result = json.load(urllib.request.urlopen(url)) # Get the API response.
        except urllib.error.URLError:
            count_of_failed_urlaccess += 1
            pass  # Fall through to the retry loop
        else:
            try:
                result_district = result['addressResult']['addresses'][0]['council']
                resultx = result['addressResult']['addresses'][0]['addressPoint']['centreX']
                resulty = result['addressResult']['addresses'][0]['addressPoint']['centreY']
                i.append(result_district)
                i.append(resultx)
                i.append(resulty)
            except KeyError:
                count_of_failed_data_retrieval += 1
            break

        if current_delay > max_delay:
            raise Exception("Too many retry attempts.")

        time.sleep(current_delay)
        current_delay *= 2  # Increase the delay each time we retry.

print('Time elapsed (hh:mm:ss.ms) {}'.format(datetime.now() - start_time))

Time elapsed (hh:mm:ss.ms) 0:00:35.077087


In [165]:
# Assess the number of properties which did not return geocoding data:

listofnonres = [x for x in properties_list if len(x) == 6]
print('Number of properties which did not return geocording data is: ')
len(listofnonres)

## Storing the geocoding data in MySQL database

Here we will load the new data to MySQL as a new table

In [171]:
properties_list_geo = properties_list.values.tolist() #Reducing the amount of data we want to upload to SQL databae to only the new geocoding data (+ id column)

In [217]:
#Picking out only the new geocoding data + the cleaned street name, and id (for reference)

dataset_onlygeocoded = [] 
for x in properties_list_geo:
    indexes = [0,5,6,7,8] 
    pickout = [x[y] for y in indexes]
    dataset_onlygeocoded.append(pickout)

In [222]:
df_withgeocoding = pd.DataFrame(dataset_onlygeocoded) #creating a pandas dataframe that is used a medium to transfer our dataset to MySQL

In [221]:
import sqlalchemy 

In [225]:
engine = sqlalchemy.create_engine('mysql+pymysql://root:password@hostname/databasename') #This allows python to interact with your MySQL database. # Tutorial to be watched to understand this line of code: https://www.youtube.com/watch?v=M-4EpNdlSuY

In [231]:
column_names2 = ['id', 'cleaned_street_name', 'district', 'longitude', 'latitude']

In [232]:
df_withgeocoding.columns = column_names2 #setting up the column names in the python dataframe so the dataset will be accepted by MySQL. 

#### Create a new table in the MySQL database and load the data:

In [ ]:
# Enter the following SQL query in MySQL Workbench:

''' 

CREATE TABLE `testdatabase`.`geocoding_data` (
  `id` INT NOT NULL,
  `cleaned_street_name` VARCHAR(150) NULL,
  `district` VARCHAR(100) NULL,
  `longitude` VARCHAR(100) NULL,
  `latitude` VARCHAR(100) NULL,
  PRIMARY KEY (`id`),
  UNIQUE INDEX `id_UNIQUE` (`id` ASC) VISIBLE);

In [233]:
#Store the geocoding data in the new table:

df_withgeocoding.to_sql(name='geocoding_data', index=False, con=engine, if_exists='append')